Workaround for pydantic v2

xl0 opened the pull request:
It's a bit heavier than expected to make mypy happy on both pydantic versions.

mesozoic commented on pyairtable/models/_base.py:
```diff
@@ -86,7 +95,8 @@ def deleted(self) -> bool:
         # Prevents implementers from changing values on readonly or non-writable fields.
-        if name in self.__class__.__fields__:
+        # Mypy can't tell that we are using pydantic v1.
+        if name in self.__class__.__fields__:  # type: ignore[operator, unused-ignore]
```
I was able to drop the `unused-ignore` portion of this without reintroducing any failures.

xl0 replied:
Are you sure you tried with v2? I think v2's BaseModel does not have __fields__, so mypy complains here.

mesozoic replied:
🤷  I'm running `tox -e mypy` and it's using the latest pydantic version, but I'm not sure it's worth our time to keep tweaking this particular nuance. Having `# type: ignore` feels kinda icky regardless, so who cares which options it needs? 😁

xl0 replied:
Yeah, it needs ignore-unused for v1 because the warning only occurs with pydantic v2.

mesozoic commented on tests/test_api_types.py:
```diff
@@ -1,5 +1,15 @@
 import pytest
-from pydantic import ValidationError
+
+if TYPE_CHECKING:  # mypy really does not like this conditional import.
+    import pydantic
```
Should this be `import pydantic.v1 as pydantic`, since mypy is going to run in a venv with `pydantic>=2`?

I'm not sure it winds up making a difference because we now have `# type: ignore` everywhere, but it seems "more correct", and perhaps it will make a difference at some point in the future.

xl0 replied:
This would cause a mypy error for v1 users, would it not?

mesozoic replied:
Hmm, I was (selfishly) thinking that since we run `tox -e mypy` in a separate environment, it will always use the latest pydantic. But you're right, we also have to think about library users who are running mypy on their own projects (and might have pinned dependencies to pydantic v1 _or_ v2).

I'm uncertain of the best way to handle this, but we probably want to test this branch as a dependency within some live code (that uses mypy) before we merge it. I can try to get to that in a week or so, but if you can sooner, all the better.

mesozoic replied:
I used this branch in a project of mine with both `pydantic~=1` and `pydantic~=2`, and the test suite (including mypy) passed in both cases. 🎉

mesozoic commented on pyairtable/api/types.py:
```diff
@@ -3,9 +3,17 @@
-from typing import Any, Dict, List, Optional, Type, TypeVar, Union, cast
+from typing import TYPE_CHECKING, Any, Dict, List, Optional, Type, TypeVar, Union, cast
+
+if TYPE_CHECKING:  # mypy really does not like this conditional import.
+    import pydantic
+else:
+    # Pydantic v2 broke a bunch of stuff. Luckily they provide a built-in v1.
+    try:
+        import pydantic.v1 as pydantic
+    except ImportError:
+        import pydantic
```
Should we move this somewhere (maybe `pyairtable.utils` or a new module `pyairtable.compat`) so it only happens once, and then import from there in the rest of our code?

xl0 replied:
I think this is a natural reaction to seeing duplicated code, but I feel in this case, the abstraction will just make it harder to follow what's going on.

xl0 replied:
Actually, since you are adding more pedantic, it definitely makes sense to move this into one place. I will do it.

mesozoic replied:
Thanks! I think most library users are not going to peek under the hood at how we're using pydantic; they're just going to want to import `Base` and `Table` and get on with life. So if it makes it easier for us (and I think it will, given how many places I'm starting to use pydantic) it's worth the time.

mesozoic reviewed (changes requested):
Fantastic! 🙌  Thank you for figuring this out! 

I like this approach because it is agnostic which version our library users require for other parts of their project. Also, since it does not require changing any of our APIs or behaviors, we can make it a patch release instead of a minor version bump.

I suggested a couple changes below. I also think we could add a few lines to `tox.ini` to ensure that our test suite runs successfully against both major versions of pydantic:

```diff
diff --git a/tox.ini b/tox.ini
index 584ae4d..142db2e 100644
--- a/tox.ini
+++ b/tox.ini
@@ -3,6 +3,7 @@ envlist =
     pre-commit
     mypy
     py3{8,9,10,11}-requests{min,max}
+    py3{8,9,10,11}-pydantic1
     coverage
 
 [gh-actions]
@@ -29,6 +30,7 @@ deps =
     -r requirements-test.txt
     requestsmin: requests==2.22.0  # Keep in sync with setup.cfg
     requestsmax: requests>=2.22.0  # Keep in sync with setup.cfg
+    pydantic1: pydantic<2  # Lots of projects still use 1.x
 
 [testenv:coverage]
 passenv = COVERAGE_FORMAT
```

I'll be very happy to merge this branch once we've got test coverage and have at least tried those other ideas.

xl0 commented:
Agree re tox, I'll add it later today.

xl0 commented:
@mesozoic I updated the PR to
- Add -pydantic1 to tox tests. Also added mypy-pydantic1.
- Moves pydantic stuff into pyairtable/pydantic.py. Can't be in utils due to circular dependency.

xl0 commented:
@mesozoic , any idea what's up with the docs failing?

mesozoic commented:
> @mesozoic , any idea what's up with the docs failing?

I think you might've forgotten to `git add` the new file. Can we call it `_compat.py`, though, to (1) make it clear that it's internal and not part of the public API, and (2) to avoid any confusion arising from the module and its contents having the same name?

xl0 commented:
@mesozoic , heh, my bad. Ok, renamed it to _compat.py.

mesozoic commented:
@xl0 One more request, could you rebase the branch with [signed commits](https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification)? Easiest way is with SSH, in my experience, but the repo requires it before we merge. I could rebase and merge it on your behalf, but then you won't get any credit. 😁

xl0 commented:
Ok, I'll check it later today.

xl0 commented:
@mesozoic , done

mesozoic commented:
Thanks very much for solving this!
